[display] Sleep / Wakeup power saving

jpeletier opened the pull request:
# What does this implement/fix?

Adds `display.sleep` and `display.wakeup` actions for putting supported display controllers into low-power mode and
waking them again.

Display drivers now publish sleep capability metadata. Configuration validation rejects these actions for unsupported
drivers and models. MIPI RGB validation also rejects configurations where the SPI control interface is unavailable or
an RGB/control pin is shared and cannot be reused to send sleep commands.

Sleep support is enabled for the hardware-tested `ILI9341`, `JC4827W543`, and `GUITION-4848S040` models.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] New developer-facing feature (adds functionality for component developers; no end-user configuration change)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — [policy](https://developers.esphome.io/contributing/code/#what-constitutes-a-c-breaking-change)
- [ ] Developer breaking change (an API change that could break external components) — [policy](https://developers.esphome.io/contributing/code/#what-is-considered-public-c-api)
- [ ] Undocumented C++ API change (removal or change of undocumented public methods that lambda users may depend on) — [policy](https://developers.esphome.io/contributing/code/#c-user-expectations)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- N/A

**Pull request in [esphome.io](https://github.com/esphome/esphome.io) with documentation (if applicable):**

- esphome/esphome.io#7104

**Pull request in [developers.esphome.io](https://github.com/esphome/developers.esphome.io) with developer documentation (if applicable):**

- N/A

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040/RP2350
- [ ] BK72xx
- [ ] RTL87xx
- [ ] LN882x
- [ ] nRF52840

Tested on `JC4827W543C` hardware. The display entered sleep and restored its previous content after wake while the
backlight remained on.

## Example entry for `config.yaml`:

```yaml
lvgl:
  on_idle:
    timeout: 30s
    then:
      - light.turn_off: display_backlight
      - lvgl.pause:
      - delay: 1s
      - display.sleep: my_display

touchscreen:
  - on_release:
      - if:
          condition:
            lvgl.is_paused:
          then:
            - display.wakeup: my_display
            - lvgl.resume:
            - light.turn_on: display_backlight
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome.io](https://github.com/esphome/esphome.io).

jpeletier committed:
define sleep/wakeup interface and implement for some displays

jpeletier committed:
define sleep/wakeup interface and implement for some displays

clydebarrow commented on esphome/components/display/display.h:
```diff
@@ -845,6 +851,26 @@ template<typename... Ts> class DisplayPageShowPrevAction : public Action<Ts...>
 };
 
+template<typename... Ts> class DisplaySleepAction : public Action<Ts...> {
+ public:
```
No need for custom classes, just use `StatelessLambdaAction` - see `components/logger/__init__.py` for an example.

jpeletier replied:
refactored!

clydebarrow commented on esphome/components/qspi_dbi/qspi_dbi.h:
```diff
@@ -16,6 +16,7 @@ namespace qspi_dbi {
 
 constexpr static const char *const TAG = "display.qspi_dbi";
 static const uint8_t SW_RESET_CMD = 0x01;
+static const uint8_t SLEEP_IN = 0x10;
```
Don't worry about updating this driver or st7701s, they are deprecated.

jpeletier replied:
Reverted changes in the deprecated components.
In exchange I implemented sleep/wakeup for mipi_spi and moved off myself from QspiDbi and it works.

clydebarrow commented on esphome/components/mipi_rgb/mipi_rgb.cpp:
```diff
@@ -101,7 +113,6 @@ void MipiRgbSpi::write_init_sequence_() {
         delay(10);
     }
   }
-  // this->spi_teardown();  // SPI not needed after this
```
We will need this retained (and uncommented) since some displays use the same pins for SPI as for the display bus. There will need to be a property to disable the teardown, which could be set by config option and/or by the sleep/wakeup actions during code generation.

jpeletier replied:
I have uncommented this line, which now tearsdown spi after init.

To simplify for now and to avoid going out of scope in this PR, I have resorted to the strategy that is working for me in production, which is to re-initialize SPI, turn off the display and tear down SPI again.

Since sleep/wakeup are not so frequent it shouldn't be much of a problem.

If later we add support for other display functions, we'll probably need to make MipiRgbSpi a template class with a SPI_TEARDOWN template boolean or something similar.

jpeletier committed:
revert changes to deprecated drivers

jpeletier committed:
revert changes to deprecated drivers

jpeletier committed:
reimplement sleep/wakeup for mipirgbspi considering spi is down after setup

jpeletier committed:
reimplement sleep/wakeup for mipirgbspi considering spi is down after setup

jpeletier committed:
implement sleep/wakeup for mipispi

jpeletier committed:
implement sleep/wakeup for mipispi

jpeletier committed:
refactor actions

jpeletier committed:
refactor actions

jpeletier committed:
Merge branch 'dev' into display-sleep

jpeletier committed:
Merge branch 'dev' into display-sleep

clydebarrow commented on esphome/components/display/display.h:
```diff
@@ -845,6 +851,26 @@ template<typename... Ts> class DisplayPageShowPrevAction : public Action<Ts...>
+
+ protected:
+  Display *display_;
+};
```
These are redundant now.

jpeletier replied:
fixed, thanks

clydebarrow commented on esphome/components/mipi_rgb/mipi_rgb.cpp:
```diff
@@ -41,6 +41,22 @@ void MipiRgbSpi::setup() {
+
+void MipiRgbSpi::wakeup() {
+  this->spi_setup();
+  this->write_command_(SLEEP_OUT);
```
You need a delay after SLEEP_OUT, 5ms minimum per the MIPI spec, 10ms used elsewhere, before sending any other commands.

jpeletier replied:
fixed, thanks

clydebarrow commented on esphome/components/mipi_spi/mipi_spi.h:
```diff
@@ -81,6 +82,11 @@ class MipiSpi : public display::Display,
   void update() override { this->stop_poller(); }
+  void sleep() override { this->write_command_(SLEEP_IN); };
+  void wakeup() override {
+    this->write_command_(SLEEP_OUT);
```
Same here, delay needed after sleep out.

jpeletier replied:
fixed, thanks

clydebarrow committed:
Merge branch 'dev' into display-sleep

jpeletier committed:
add delays

jpeletier committed:
add delays

jpeletier committed:
remove redundant actions

jpeletier committed:
remove redundant actions

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

clydebarrow commented on esphome/components/mipi_rgb/mipi_rgb.cpp:
```diff
@@ -41,6 +41,23 @@ void MipiRgbSpi::setup() {
+  delay(10);
+  this->write_command_(DISPLAY_ON);
+  this->spi_teardown();
+  delay(10);
```
I don't think this delay is required.

jpeletier replied:
Removed. Both trailing `delay(10)` calls (after `spi_teardown()` in `sleep()` and `wakeup()`) served no purpose since no command follows them; kept the required settle delay between `SLEEP_OUT` and `DISPLAY_ON`. Verified on GUITION-4848S040 hardware that sleep/wakeup still work correctly.

jpeletier commented:
> The sleep actions should only be available if the driver python code sets a flag to say so - this prevents it being used on drivers that have not been updated.

How do I create actions that only work for a subset of drivers of the same class? I tried by creating a fake class like this:

```python
# in components/display/__init__.py

# ...
Display = display_ns.class_("Display", cg.PollingComponent)
SleepTrait = display_ns.class_("SleepTrait")
# ...
DISPLAY_SLEEP_WAKEUP_ACTION_SCHEMA = maybe_simple_id(
    {
        cv.GenerateID(CONF_ID): cv.All(cv.use_id(Display), cv.use_id(SleepTrait)),
    }
)
# ...
```

Then, in some display driver:

```python
my_driver_ns = cg.esphome_ns.namespace("my_driver")
my_driver = mipi_rgb_ns.class_(
    "MyDriver", display.Display, cg.Component, display.SleepTrait
)
```
This works to enforce the constraint, but the error message you get when you use `display.sleep` on a non-supported display looks as follows:

```
- display.sleep: 
    
    ID 'device_display' of type my_driver::MyDriver doesn't inherit from display::SleepTrait. 
    Please double check your ID is pointing to the correct value.
```
`display::SleepTrait` doesn't actually exist as a C++ class, so the error is misleading.

Am I the first one to be writing a set of actions that can only work for certain devices? Could you point me to an example in some other component that already does this?.

I think it would be much easier to just let C++ do the job and show an error message at runtime if the driver does not support the action.

We could have a mixin class we can add to drivers that support a similar way of sleeping, namely:

```cpp
// C++ concept to constrain templates so these are mixin templates for displays only.
template <typename T>
concept IsDisplay = std::derived_from<T, display::Display>;

// define the way of implementing sleep for MipiRgb displays:
template <IsDisplay BaseDisplay>
class MipiRgbSleepTrait : public BaseDisplay {
 public:
  void sleep() override { /* sleep code */ };
  void wakeup() override { /* wakeup code */ };
```
Then, for example, `GUITION-4848S040` would be
```cpp
class GUITION_4848S040 : public MipiRgbSleepTrait<MipRgbSpi> {
    using MipiRgbSpi::MipiRgbSpi;
   // handle other quirks here, perhaps some pin needs to be HIGH before sleeping
   // or wakeup requires further steps compared to most MipiRgb displays
   void wakeup() override { 
      MipiRgbSleepTrait::wakeup(); // do normal wakeup
      /* special extra wakeup sequence for this display, or extra timing */
   };
   
   /* other quirks */
   void clear() override {
     Display::clear();
     /* other stuff for this board to work */
   }
  
};
```
The python file in "models" would reference a quirks class if needed for a particular board:
```python

st7701s.extend(
    "GUITION-4848S040",
    width=480,
    height=480,
    data_rate="2MHz",
    #...
    custom_class = "GUITION_4848S040", # new parameter
)
```
This way, each model can potentially have its own minimal class in case it needs to deal with minor differences instead of having to mess with python code. It also makes it more resilient at c++ level, so that if someone calls a sleep method from a lambda when it was forbidden at validation time, python wouldn't detect it but c++ would certainly do.

jpeletier commented:
I've been looking at how to do this.

> The sleep actions should only be available if the driver python code sets a flag to say so - this prevents it being used on drivers that have not been updated.

How do I create actions that only work for a subset of drivers of the same class? I tried by creating a fake class like this:

```python
# in components/display/__init__.py

# ...
Display = display_ns.class_("Display", cg.PollingComponent)
SleepTrait = display_ns.class_("SleepTrait")
# ...
DISPLAY_SLEEP_WAKEUP_ACTION_SCHEMA = maybe_simple_id(
    {
        cv.GenerateID(CONF_ID): cv.All(cv.use_id(Display), cv.use_id(SleepTrait)),
    }
)
# ...
```

Then, in some display driver:

```python
my_driver_ns = cg.esphome_ns.namespace("my_driver")
my_driver = mipi_rgb_ns.class_(
    "MyDriver", display.Display, cg.Component, display.SleepTrait
)
```
This works to enforce the constraint, but the error message you get when you use `display.sleep` on a non-supported display looks as follows:

```
- display.sleep: 
    
    ID 'device_display' of type my_driver::MyDriver doesn't inherit from display::SleepTrait. 
    Please double check your ID is pointing to the correct value.
```
`display::SleepTrait` doesn't actually exist as a C++ class, so the error is misleading.

Am I the first one to be writing a set of actions that can only work for certain devices? Could you point me to an example in some other component that already does this?.

I think it would be much easier to just let C++ do the job and show an error message at runtime if the driver does not support the action.

We could have a mixin class we can add to drivers that support a similar way of sleeping, namely:

```cpp
// C++ concept to constrain templates so these are mixin templates for displays only.
template <typename T>
concept IsDisplay = std::derived_from<T, display::Display>;

// define the way of implementing sleep for MipiRgb displays:
template <IsDisplay BaseDisplay>
class MipiRgbSleepTrait : public BaseDisplay {
 public:
  void sleep() override { /* sleep code */ };
  void wakeup() override { /* wakeup code */ };
```
Then, for example, `GUITION-4848S040` would be
```cpp
class GUITION_4848S040 : public MipiRgbSleepTrait<MipRgbSpi> {
    using MipiRgbSpi::MipiRgbSpi;
   // handle other quirks here, perhaps some pin needs to be HIGH before sleeping
   // or wakeup requires further steps compared to most MipiRgb displays
   void wakeup() override { 
      MipiRgbSleepTrait::wakeup(); // do normal wakeup
      /* special extra wakeup sequence for this display, or extra timing */
   };
   
   /* other quirks */
   void clear() override {
     /* other stuff for this board to work */
   }
  
};
```
The python file in "models" would reference a quirks class if needed for a particular board:
```python

st7701s.extend(
    "GUITION-4848S040",
    width=480,
    height=480,
    data_rate="2MHz",
    #...
    custom_class = "GUITION_4848S040", # new parameter
)
```
This way, each model can potentially have its own minimal class in case it needs to deal with minor differences instead of having to mess with python code. It also makes it more resilient at c++ level, so that if someone calls a sleep method from a lambda when it was forbidden at validation time, python wouldn't detect it but c++ would certainly do.

jpeletier committed:
[display] Validate sleep and wakeup support

jpeletier committed:
[mipi_rgb] Defer sleep pin-sharing check to final validation

The shared-pin detection read PIN_SCHEMA_REGISTRY.get_count() during
CONFIG_SCHEMA, where pin-usage counts are only partially populated. Whether a
co-user of a GPIO had registered yet depended on component validation order, so
a shared pin could be missed and display.sleep wrongly permitted.

Store the pin-sharing test as a deferred callable in the display metadata and
evaluate it from display's final-validate step, where the pin registry is
complete regardless of component order.

jpeletier committed:
[mipi_rgb] Remove unnecessary delays after SPI teardown in sleep/wakeup

The trailing delay(10) calls in sleep() and wakeup() ran after spi_teardown()
with no command following them, so they only blocked the main loop without
affecting the panel. Keep the required settle delay between SLEEP_OUT and the
next command. Verified on GUITION-4848S040 hardware: sleep and wakeup still
work correctly.

jpeletier committed:
[mipi_spi] Add settle delay after SLEEP_IN in sleep()

The controller spec requires waiting >=5ms after SLEEP_IN before the next
command. Unlike mipi_rgb (SPI-idle between sleep and wake), MipiSpiBuffer keeps
its poller running and issues SPI writes on the next update(), which can land
within a few ms of SLEEP_IN. Add a 5ms settle delay to honor the spec.

jpeletier committed:
[mipi_rgb] Simplify sleep capability detection using allow_other_uses

Detect a shared RGB/control pin by reading allow_other_uses on the display's own
pin config instead of PIN_SCHEMA_REGISTRY.get_count(). A pin used in more than one
place must set allow_other_uses (enforced both ways in pins.py), so the flag is an
order-independent, order-safe signal available already at CONFIG_SCHEMA time.

This removes the deferred callable that the previous commit stored in the display
metadata to work around get_count() only being complete at final validation. The
DisplayMetaData sleep fields collapse back to two plain values (supports_sleep,
sleep_unsupported_reason); the Callable and CORE imports are no longer needed.

Note: in CORE.testing_mode pins.py skips reuse enforcement, so a colliding pin need
not set allow_other_uses; the check would then report the display as sleep-capable.
Harmless -- those synthetic grouped configs don't exercise sleep actions.

jpeletier committed:
[display] Store sleep support as a single reason field

Replace the two-field (supports_sleep bool + sleep_unsupported_reason) metadata
with one source of truth: sleep_unsupported_reason, where None means supported and
any string is the reason sleep is unavailable (defaulting to a generic 'not
implemented' message). supports_sleep becomes a derived read-only property, so the
contradictory state (supports_sleep=True with a reason set) is no longer
representable.

add_metadata keeps its convenient supports_sleep=/sleep_unsupported_reason=
signature and derives the stored value, so display drivers and existing tests need
no changes.

jpeletier committed:
[mipi_rgb] Add settle delay after SLEEP_IN
